Generic codecov submissions to support enterprise installations

ccoffrin opened the pull request:
I am fairly new to Julia so I might have done this all wrong, please don't hesitate to tell me so.  :-)

I ran into a problem with the current Codecov implementation where I could not submit reports to an enterprise installation of Codecov.  The key problem was that the uri needs to point to something like "enterprise-codecov.com" instead of "codecov.io", and even if I am using Travis-CI I need to provide a repo token.

Summary of the proposed changes:
- adding a new generic submission method (i.e. `generic_submit`)
- re-casting existing submission methods in terms of `generic_submit`
- deprecating `submit_token` in favor of a hopefully more intuitive name `submit_local`
- making environment variables consistent with [codecov-bash](https://github.com/codecov/codecov-bash) 
- deprecating `REPO_TOKEN` in favor of `CODECOV_TOKEN` used by codecov-bash

The approach:
To fix my enterprise codecov issues, and hopefully guard against similar issues in the future, I took the approach of making a generic_submit function, which builds the codecov uri from the function's keyword arguments.  Hence, the upload uri can be adapted or extended easily by passing new keyword arguments this function.

With the new generic_submit, we can now see submit and submit_token as simply defining some default values to pass to generic_submit.  Any of these default values can be overridden simply by passing the appropriate keyword argument to those functions.  When no default value or keyword argument are specified, environment variables are checked and used if available.

I have verified this works on enterprise use cases, but have not tested the non-enterprise uses cases.

tkelman commented on src/codecovio.jl:
```diff
@@ -40,53 +40,129 @@ module Codecov
+            end
+        end
+        @assert codecov_url != "None"
+        @assert codecov_url[end] != "/"
```
should these give better error messages if people use this incorrectly?

tkelman commented on src/codecovio.jl:
```diff
@@ -40,53 +40,129 @@ module Codecov
+    function set_defaults(args_array; kwargs...)
+        defined_names = [k for (k,v) in args_array]
+        for kwarg in kwargs
+            if ~(kwarg[1] in defined_names)
```
`~` is bitwise, usually `!` is preferred for boolean not

tkelman commented on src/codecovio.jl:
```diff
@@ -40,53 +40,129 @@ module Codecov
+
+    make depreciated
+    """
+    function submit_token(fcs::Vector{FileCoverage})
```
there's an `@deprecate` macro you can use

ccoffrin commented:
In these commits I,
- addressed the minor suggestions of @tkelman 
- made the default codecov_url value easier to override
- added tests to ensure the urls are being generated correctly

I not 100% sure I used the @depricated macro correctly, @tkelman can you have a look at line 94?

tkelman commented on src/codecovio.jl:
```diff
@@ -40,53 +40,138 @@ module Codecov
+        submit_local(fcs; kwargs...)
+    end
+
+    @deprecate submit_token submit_local
```
using the `@deprecate` macro should allow you to leave out the placeholder definition of `submit_token` above

tkelman commented on test/runtests.jl:
```diff
@@ -61,3 +61,196 @@ cd(Pkg.dir("Coverage")) do
+# clear system defined ENV vars in preparation for testing
+
+delete!(ENV, "CODECOV_URL")
+delete!(ENV, "CODECOV_TOKEN")
```
would probably be better to use `withenv` for temporarily clearing or setting these and automatically restoring them

tkelman commented on src/codecovio.jl:
```diff
@@ -40,53 +40,138 @@ module Codecov
+    values match the Codecov upload/v2 API specification.  
+    The `codecov_url` keyword argument or the CODECOV_URL environment variable 
+    can be used to specify the base path of the uri.
+    The `julia_test` keyword can be used to prevent the http request from 
```
should this be called something different? `dry_run` maybe?

ccoffrin commented:
In these commits I integrated the suggestions of @tkelman.  @depricated now seems to be working as expected.

ccoffrin commented on test/runtests.jl:
```diff
@@ -61,3 +61,193 @@ cd(Pkg.dir("Coverage")) do
+    @test !contains(codecov_url, "service")
+
+    # default values in depreciated call
+    codecov_url = extract_codecov_url( () -> Coverage.Codecov.submit_token(fcs) )
```
@tkelman is there a standard/simple way that tests can check for a warning?

This [thread](https://groups.google.com/forum/#!msg/julia-users/X90Tvh4kixU/DNbB-wJFacAJ) suggests there is not a simple way to do it.

ccoffrin commented:
Are there any more concerns with these changes?

timholy commented:
On a brief scan this looks fine to me: it seems very thorough, and it passes tests on julia-0.4 (on 0.5, there's an unrelated error). Since you put some time into reviewing this @tkelman, I'll ask: OK if I merge this?

ccoffrin commented:
Thanks a bunch!
